java/ql/lib/semmle/code/java/dataflow/TypeFlow.qll

```diff
@@ -413,30 +429,41 @@ private predicate downcastSuccessor(VarAccess va, RefType t) {
   )
 }
 
+private Expr getAProbableAlias(Expr e) {
```

Comment:
This refactor is inherently quadratic, so this looks dangerous. The proper fix here is to add (maybe just a tiny bit of) value numbering. It could still be private and only cover these two cases, but it should be a def-use relation rather than a use-use cartesian product.